nevergrad/optimization/optimizerlib.py

```diff
@@ -1379,30 +1379,50 @@ class cGA(base.Optimizer):
-        self._penalize_cheap_violations = False  # Not sure this is the optimal decision.
-        self.p: np.ndarray = np.ones((1, self.dimension)) / num_categories
-        self.llambda = 2 * (self.budget if self.budget is not None else max(num_workers, 40))
-        self._value_candidate: Optional[Tuple[float, np.ndarray]] = None
+        if arity is None:
+            arity = len(instrumentation.possibilities) if hasattr(instrumentation, "possibilities") else 2  # type: ignore
+        self._arity = arity
+        self._penalize_cheap_violations = True  # Not sure this is the optimal decision.
+        # self.p[i][j] is the probability that the ith variable has value 0<=j< arity.
+        self.p: np.ndarray = np.ones((self.dimension, arity)) / arity
+        # Probability increments are of order 1./self.llambda
+        # and lower bounded by something of order 1./self.llambda.
+        self.llambda = 2 * (self.budget if self.budget is not None else max(num_workers, 40))  # FIXME: no good heuristic ?
+        # CGA generates a candidate, then a second candidate;
+        # then updates depending on the comparison with the first one. We therefore have to store the previous candidate.
+        self._previous_value_candidate: Optional[Tuple[float, np.ndarray]] = None
 
     def _internal_ask_candidate(self) -> base.Candidate:
-        unif = self._rng.uniform(size=self.dimension)
-        data = (unif > 1 - self.p[0]).astype(float)
+        # Multinomial.
+        values: List[int] = []
+        for i, w in enumerate(self.p):
+            values += [sum(self._rng.uniform() > np.cumsum(w))]
+        data = inst.discretization.noisy_inverse_threshold_discretization(values, arity=self._arity, gen=self._rng)
         return self.create_candidate.from_data(data)
 
     def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> None:
-        if self._value_candidate is None:
-            self._value_candidate = (value, candidate.data)
+        if self._previous_value_candidate is None:
+            self._previous_value_candidate = (value, candidate.data)
         else:
-            winner, loser = self._value_candidate[1], candidate.data
-            if self._value_candidate[0] > value:
+            winner, loser = self._previous_value_candidate[1], candidate.data
+            if self._previous_value_candidate[0] > value:
                 winner, loser = loser, winner
-
-            self.p[0] = self.p[0] + (winner != loser) * (2 * winner - 1) / self.llambda
-            self.p[0] = np.clip(self.p[0], 0, 1)
-            self._value_candidate = None
+            winner_data: ArrayLike = [int(d) for d in winner.data]
+            loser_data: ArrayLike = [int(d) for d in loser.data]
+            winner_data = [float(d) for d in inst.discretization.threshold_discretization(winner_data ,arity=self._arity)]
+            loser_data = [float(d) for d in inst.discretization.threshold_discretization(loser_data ,arity=self._arity)]
```

Comment:
Why not just `inst.discretization.threshold_discretization(winner_data ,arity=self._arity)` ? 
You don't need to transform them as floats, especially because your comparison later 
`winner_data[i] != loser_data[i]` relies on the fact that these floats are actually integers.
You also wouldn't need to round them on ll1422, 1423.

Reply:
done